src/LibraryViewExtensionMSWebBrowser/Handlers/NodeItemDataProvider.cs

```diff
@@ -0,0 +1,177 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Dynamo.Search;
+using Dynamo.Search.SearchElements;
+using Dynamo.LibraryViewExtensionMSWebBrowser.Handlers;
+using Newtonsoft.Json;
+
+// TODO there are many clases in this file which are duplicates in LibraryViewExtension - 
+// these can be refactored out into a shared core.
+
+namespace Dynamo.LibraryViewExtensionMSWebBrowser
+{
+    class NodeData
+    {
+        public string creationName { get; set; }
+        public string module { get; set; }
+    }
+
+    class LoadedTypeItem
+    {
+        public string fullyQualifiedName { get; set; }
+        public string iconUrl { get; set; }
+        public string contextData { get; set; }
+        public string parameters { get; set; }
+        public string itemType { get; set; }
+        public string description { get; set; }
+        public string keywords { get; set; }
+    }
+
+    class LoadedTypeData<T> where T : LoadedTypeItem
+    {
+        public List<T> loadedTypes { get; set; }
+    }
+
+    /// <summary>
+    /// Provides json resource data for all the loaded nodes
+    /// </summary>
+    class NodeItemDataProvider : ResourceProviderBase
+    {
+        protected NodeSearchModel model;
+        private IconResourceProvider iconProvider;
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="items"></param>
+        public NodeItemDataProvider(NodeSearchModel model) : base(false)
+        {
+            this.model = model;
+        }
+
+        public NodeItemDataProvider(NodeSearchModel model, IconResourceProvider iconProvider)
+        {
+            this.model = model;
+            this.iconProvider = iconProvider;
+        }
+
+        public override Stream GetResource(string url, out string extension)
+        {
+            extension = "json";
+            //pass false to keep original icon urls
+            return GetNodeItemDataStream(model.SearchEntries, false);
+        }
+
+        protected Stream GetNodeItemDataStream(IEnumerable<NodeSearchElement> searchEntries, bool replaceIconURLWithData)
+        {
+            var ms = new MemoryStream();
+            var sw = new StreamWriter(ms);
+            var serializer = new JsonSerializer();
+            var stringBuilder = new StringBuilder();
+            var data = CreateObjectForSerialization(searchEntries);
+
+            if (replaceIconURLWithData)
```

Comment:
we have the option of replacing the nodeItems iconUrl property with the base64Data before it gets sent to the library UI - or not (and letting) the image resolve method handle it - for searching, because the results are quite small we embed the icons as it will be fast.

Reply:
Excellent comment, would you put this in the code. I feel this explains it well

Reply:
will do